server/src/main/java/org/elasticsearch/index/snapshots/IndexShardSnapshotStatus.java

```diff
@@ -78,6 +95,7 @@ public enum AbortStatus {
     private long processedSize;
     private String failure;
     private final SubscribableListener<AbortStatus> abortListeners = new SubscribableListener<>();
+    private volatile AbortStatus abortStatus = AbortStatus.NOT_ABORTED;
```

Comment:
Do we need this extra field? It seems to me that the `stage` field represents the long term overall state and abortStatus is more transient and only needed for the listener to make decision. I wonder whether we can buff up the `AbortedSnapshotException` class to carry the abortStatus and avoid to manage an extra state? But maybe you have future plans on how to use it?

Reply:
I considered that too, but we end up needing to store this here either way because we generate the `AbortedSnapshotException` later in places like `ensureNotAborted()`.

Reply:
Fair enough. 
Mostly curious: Other fields, e.g. `stage`, ues `AtomicReference`, while you chose `volatile` for this new field. Could there be any reason that we might need the extra safety from the AtomicReference?

Reply:
I don't think there's any more safety from an `AtomicReference`, just extra features (notably the compare-and-set operation). It doesn't look like it's needed for `shardSnapshotResult` or `generation`.